web-search-engines/langchain4j-web-search-engine-searchapi/src/main/java/dev/langchain4j/web/search/searchapi/SearchApiWebSearchEngine.java

```diff
@@ -0,0 +1,277 @@
+			final Boolean logRequests, final Consumer<Map<String, Object>> customizeParametersFunc) {
+
+		this.apiKey = ensureNotBlank(apiKey, "apiKey");
+		this.engine = getOrDefault(engine, DEFAULT_ENGINE);
+		this.logRequests = getOrDefault(logRequests, false);
+		this.customizeParametersFunc = customizeParametersFunc;
+
+		this.searchapiClient = SearchApiClient.builder().baseUrl(DEFAULT_BASE_URL).apiKey(this.apiKey)
+				.engine(this.engine).timeout(getOrDefault(timeout, ofSeconds(30))).logRequests(this.logRequests)
+				.build();
+	}
+	
+	public SearchApiWebSearchEngine(final String apiKey, final String engine) {
+		this(apiKey, engine, null, false, null);
+	}
+	
+	public SearchApiWebSearchEngine(final String apiKey, final String engine, final Consumer<Map<String, Object>> customizeParametersFunc) {
+		this(apiKey, engine, null, false, customizeParametersFunc);
+	}	
+
+	@Override
+	public WebSearchResults search(final WebSearchRequest webSearchRequest) {
+		ensureNotNull(webSearchRequest, "webSearchRequest");
+
+		final SearchApiRequest request = SearchApiRequest.builder()
+				.q(webSearchRequest.searchTerms())
+		        .safe(webSearchRequest.safeSearch())
+				.num(getOrDefault(webSearchRequest.maxResults(), DEFAULT_MAX_RESULTS)) // max results should default to 5
+				.page(webSearchRequest.startPage())
+				.build();
+		
+		// append any parameters included with a search
+		if (webSearchRequest.additionalParams() != null) {
+			request.getParams().putAll(webSearchRequest.additionalParams());
+		}
+
+		// current & future search engines must either:
+        // 1a. override this method to customize the search parameters
+        customizeSearchRequest(request, webSearchRequest);
+        
+        // 1b. or use the functional interface
+        if (customizeParametersFunc != null) {
+        	customizeParametersFunc.accept(request.getParams());
+        }
+
+        // 2. conduct the search
+        final SearchApiResponse searchapiResponse = searchapiClient.search(request);
+        
+        // 3. structure the results
+        final List<WebSearchOrganicResult> results = searchapiResponse.getOrganicResults().stream()
+				.map(result -> WebSearchOrganicResult.from(result.getTitle(), URI.create(result.getLink()),
+						getOrDefault(result.getSnippet(), ""), // for the first few runs of the query I tried, "snippet" was null
+						null, // searchapi does not return "content"
+						toResultMetadataMap(result)))
+				.collect(toList());
+
+		// 4. if the search query has a direct answer, append it to the top of the organic results
+		customizeSearchResults(searchapiResponse, results);
+
+		final Double pageNumber = searchapiResponse.getPagination().containsKey("current") ? 
+				(Double)searchapiResponse.getPagination().get("current") : 1;
+		final WebSearchInformationResult result = WebSearchInformationResult.from(
+				Long.valueOf(results.size()),
```

Comment:
the value must be `search_information.total_results`. These should be the total results that the search produces regardless of any added filtering.